### DISCUSSION TITLE
v6.4 Add the navigation state in the LoaderFunctionArgs

### DISCUSSION THREAD

**Discussion by mattiaz9** [remix-run/react-router#9137] [Proposals]
I'd like to know the state passed in the `<Link />` component inside the loader function args.

Is it possibile to have this?

**Comment by brophdawg11** [comment]
Can you explain your use-case a bit more?  Gernerally this doesn't seem like a good idea to me since your loader would no longer be [idempotent](https://developer.mozilla.org/en-US/docs/Glossary/Idempotent).  And since loaders handle GET requests, they should be idempotent by default. 

Would a query string be an option?  Or would you be better off performing a `POST` navigation and sending information in `FormData`?

**Reply by mattiaz9** [reply]
Let's suppose I load a bunch of items in the parent layout.
When I open the details view I want to pass the already downloaded state in the loader function so I can skip to download another time the data that I already have.
On the other hand when I access directly to the details page I want to download the data.

So basically it's just to optimize server requests and speed.

**Reply by devpeerapong** [reply]
Is this correct for my understanding of your problem ?

You have route `/notes` that fetch all the notes in the `loader`.
Suppose you click on the note title, It will show you the detail with `<NoteDetail noteId={id} />` component.
You don't want to make another network request to the server since all the data has already been loaded in `/notes`.
Now you cannot use `useLoaderData` because it's a child component of the route `/notes`

If so, I think it is impossible to do with React Router alone since it doesn't have caching.
You need to bring in state management to handle [server cache](https://kentcdodds.com/blog/application-state-management-with-react#server-cache-vs-ui-state). Some tools like [Redux](https://redux-toolkit.js.org/), [Tanstack Query](https://tanstack.com/query/v4/docs/reference/QueryClient) or simple `Map` should work.

But if you really don't want to add another library.
You can use [`useRouteLoaderData`](https://beta.reactrouter.com/en/dev/hooks/use-route-loader-data) to get the data in the loader for specific route (in this case `useRouteLoaderData("notes")`.
But this feels a bit too hacky since the component need to know what route its need to get data from.

**Reply by brophdawg11** [reply]
In these types of UX we generally find that you don't want _all_ the data in the list view that the detail view needs.  You want the trimmed data containing only the fields rendered in the list.  Then in the detail view you need _all_ the data.  If you go this route of trying to load everything in the list and then pass it through state, it seems like you'd be de-optimizing the parent loader (by including too much info), just to optimize a click to the child that may never happen?

For example, if your items have 10 fields on them, but you only display 2 in the list view - you don't necessarily want to send the other 8 over the wire in the parent call.  If you have 100 items you'd now be sending 800 unused fields in the parent loader.

If you really want to go this route, I would look at providing the data through [`useOutletContext`](https://reactrouter.com/docs/en/v6/hooks/use-outlet-context).  If you access the details route directly the parent loader is still going to run so the data will still be downloaded

**Reply by brophdawg11** [reply]
Ah thank you @devpeerapong for reminding me of the new `useRouteLoaderData` hook too!  That's another good option

**Reply by mattiaz9** [reply]
@brophdawg11 sometimes you can't have trimmed data because you don't have control over the server apis.
The `useRouteLoaderData` seems interesting, but it can't always work.
Let's take for instance YouTube. You click on a video and you go to the watch page. In that case you can't pass the state from one page to another, right?

**Reply by brophdawg11** [reply]
So no control over the server and no parent/child hierarchy to leverage shared data - this is getting to be a bit of an edge case ;)

It sounds like you're effectively looking for a client-side cache accessible via the loader - any reason you can't just set up your own client-local cache?  Your loaders can then short circuit and avoid a network call if you find the item in the cache.

Take a look at this example: https://codesandbox.io/s/client-side-loader-cache-tb7ut7

**Reply by mattiaz9** [reply]
Thank you @brophdawg11. That's probably the best way to implement that scenario!

**Reply by brophdawg11** [reply]
You're welcome - glad it helped!
